src/utp_stream.cpp

```diff
@@ -981,35 +1062,59 @@ void utp_socket_impl::writable()
 #endif
 	TORRENT_ASSERT(m_stalled);
 	m_stalled = false;
+	maybe_trigger_writeable_callback({});
 	if (should_delete()) return;
 
-	// this handles the case where send_fin() was called while stalled
-	if (state() == state_t::fin_sent && m_outbuf.at(m_seq_nr) == nullptr)
-		send_pkt(pkt_fin);
 	// if the socket stalled while sending an ack then there will be a
 	// pending deferred ack. make sure it gets sent out
 	else if (!m_deferred_ack || send_pkt(pkt_ack))
 		while(send_pkt());
 
-	maybe_trigger_send_callback();
+	// try to flush the last part of the send buffer before sending a FIN
+	if (m_out_eof && m_nagle_packet) send_pkt();
+
+	// this handles the case where send_fin() was called while stalled
+	if (m_out_eof
+		&& !m_nagle_packet
+		&& m_write_buffer_size == 0
+		&& state() == state_t::connected
```

Comment:
When `send_fin()` fails due to stalling, it still progresses to `state_t::fin_sent`. Should this check be `state() == state_t::fin_sent` (as it was prior to this patch) in order to pick it up?

Reply:
good point. Maybe the state shouldn't even be transitioned in case the socket is stalled, failing to send the FIN.

Reply:
it's not obvious to me that checking the socket state here is important either.

Reply:
This bit might actually be able to be dropped entirely. If sending the FIN packet stalls, it gets marked for resend and ends up in `m_outbuf`, and I think should get resent in one of the previous `send_pkt()` calls above.

It looks like the second commit in #7153 might have been a bit wonky and tried to address something that was already solved by it's first commit, so the logic here no longer seems to makes sense.

Reply:
yeah. I hesitate dropping this until there are proper unit tests for these cases.

Reply:
If I'm understanding it correctly, the old conditions would never be met, so functionally nothing was happening here anyway. When `state() == state_t::fin_sent`, `m_outbuf.at(m_seq_nr)` would be the FIN packet. Once it's ACK'd and removed from `m_outbuf`, we'll no longer be in the `fin_sent` state. The assumption was that a stalled FIN packet wouldn't have made its way into `m_outbuf` (which now happens) and could be detected here and re-sent - it instead gets resent in the `send_pkt()` calls above along with any other packets needing to be resent.

I'm not sure what effect the new conditions here using the `connected` state instead of the `fin_sent` state might have.

Reproducing stalled FIN packets in a test would be ideal.